### PR TITLE
Elements outside visible area indicators

### LINKED ISSUE #3859: Show indicator arrows for elements outside visible canvas area

### PR DESCRIPTION
**Pull request by ruggi**
Fixes #3859 

**Problem:**

If a selected/highlighted element is outside the visible canvas area it can be difficult to figure out "where" it actually is.

**Fix:**

This PR introduces indicator arrows that appear at the edges of the visible canvas, pointing to selected/highlighted elements that are outside of it.

https://github.com/concrete-utopia/utopia/assets/1081051/89b4e899-c3b1-40cf-9db3-94af72df5d63

- The arrows are oriented relative to the center of the canvas.
- If there are 2+ overlapping arrows, they get clustered and only one is displayed, with a counter at its side
- If the arrow is clicked, the canvas scrolls to the target element. Note: the scroll will bring the element to the top-left corner of the canvas instead than center, but that's besides the scope of this PR (the `scrollToElement` action is used)
- To avoid having arrows disappearing/hidden by the canvas toolbar, the canvas toolbar is considered a safe area and arrows on the left side of the canvas will be shown on the right _of the toolbar_ (if applicable)
- If there are 10+ clustered elements, the cluster label will be forced to `"10+"`

Note: to avoid having to do a very complicated set of zoom/scale/positioning adjustments, I positioned the component rendering the arrows so that it is _not_ part of the zoomed canvas (and I don't think it should be, since canvas zooming should not affect this kind of indicators) but instead relative to its container.

### CODE REVIEW

File: editor/src/components/canvas/controls/new-canvas-controls.tsx [hunk 1]
```diff
@@ -173,55 +179,67 @@ export const NewCanvasControls = React.memo((props: NewCanvasControlsProps) => {
...
+          >
+            <NewCanvasControlsInner
+              windowToCanvasPosition={props.windowToCanvasPosition}
+              localSelectedViews={localSelectedViews}
+              localHighlightedViews={localHighlightedViews}
+              setLocalSelectedViews={setSelectedViewsLocally}
+              setLocalHighlightedViews={setLocalHighlightedViews}
+            />
+          </div>
+          <ElementContextMenu contextMenuInstance='context-menu-canvas' />
```
Comment (balazsbajorics) [inline editor/src/components/canvas/controls/new-canvas-controls.tsx:239]: I think if you make ElementsOutsideVisibleAreaIndicators a sibling of ElementContextMenu it'll already be outside the zooming / scrolling part of canvas, and then you can spare the fragment wrapping here

File: editor/src/components/canvas/controls/new-canvas-controls.tsx [hunk 2]
```diff
@@ -565,3 +582,100 @@ const SelectionAreaRectangle = React.memo(
 )
 
 SelectionAreaRectangle.displayName = 'SelectionAreaRectangle'
+
+const ElementsOutsideVisibleAreaIndicators = React.memo(
```
Comment (balazsbajorics) [inline editor/src/components/canvas/controls/new-canvas-controls.tsx:586]: please move this to a dedicated file

File: editor/src/components/canvas/controls/new-canvas-controls.tsx [hunk 3]
```diff
@@ -173,55 +179,67 @@ export const NewCanvasControls = React.memo((props: NewCanvasControlsProps) => {
     [drop],
   )
 
+  const ref = React.useRef<HTMLDivElement | null>(null)
+
+  const elementsOutsideVisibleAreaIndicators = useElementsOutsideVisibleArea(
```
Comment (balazsbajorics) [inline editor/src/components/canvas/controls/new-canvas-controls.tsx:184]: I think this should be moved inside the `ElementsOutsideVisibleAreaIndicators`

File: editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+
+export type ElementOutsideVisibleAreaIndicator = {
+  id: string
+  path: ElementPath
+  position: WindowPoint
+  angle: number
+  cluster: number
+}
+
+export function useElementsOutsideVisibleArea(
```
Comment (balazsbajorics) [inline editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx:41]: this is very nice! :)

**Review by balazsbajorics — APPROVED** [review]
this is a great feature, feels very natural to see it / use it!

**Review by enidemi — APPROVED** [review]

File: editor/src/components/canvas/controls/new-canvas-controls.tsx (hunk 3 above)
Comment (ruggi) [reply editor/src/components/canvas/controls/new-canvas-controls.tsx]: https://github.com/concrete-utopia/utopia/pull/3842/commits/30a89e064c08ac48c0a1ebb9b506cdd91e68d927

File: editor/src/components/canvas/controls/new-canvas-controls.tsx (hunk 2 above)
Comment (ruggi) [reply editor/src/components/canvas/controls/new-canvas-controls.tsx]: https://github.com/concrete-utopia/utopia/pull/3842/commits/30a89e064c08ac48c0a1ebb9b506cdd91e68d927

**Review by gbalint — APPROVED** [review]

File: editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+    (store) => (store.editor.navigator.minimised ? 0 : LeftPaneDefaultWidth),
+    'useElementsOutsideVisibleArea navigatorMinimised',
+  )
+
+  const elements = React.useMemo(() => {
+    return uniqBy([...localSelectedViews, ...localHighlightedViews], EP.pathsEqual)
+  }, [localSelectedViews, localHighlightedViews])
+
+  const framesByPathString = React.useMemo(() => {
+    return elements.reduce((acc, path) => {
```
Comment (seanparsons) [inline editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx:74]: Might as well just loop over the elements if we're mutating `acc`.

File: editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+          }
+        })
+        // Group the indicators into clusters
+        .reduce((arr, indicator): ElementOutsideVisibleAreaIndicator[] => {
+          const index = arr.findIndex((other) => {
+            const distanceBetween = distance(indicator.position, other.position)
+            return distanceBetween < minClusterDistance
+          })
+          if (index >= 0) {
+            arr[index].cluster++
```
Comment (seanparsons) [inline editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx:180]: If this was a loop, then you could also eliminate the intermediary array that the above `.map` produces.

File: editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+  currentValue: number,
+  min: ElementOutsideVisibleAreaDirectionBaseValue,
+  max: ElementOutsideVisibleAreaDirectionBaseValue,
+): number {
+  if (directions.includes(min.direction)) {
+    return min.baseValue
+  } else if (directions.includes(max.direction)) {
+    return max.baseValue
+  } else {
+    return Math.max(Math.min(currentValue, max.baseValue), min.baseValue)
```
Comment (seanparsons) [inline editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx:217]: We've got a function `clamp` for this BTW.

**Review by seanparsons — APPROVED** [review]

File: editor/src/components/canvas/controls/elements-outside-visible-area.tsx [hunk 8]
```diff
@@ -0,0 +1,127 @@
+/* eslint-disable @typescript-eslint/ban-types */
```
[the remark this answers is not in the document]
Comment (ruggi) [reply editor/src/components/canvas/controls/elements-outside-visible-area.tsx]: heh, copypasta. Removed it https://github.com/concrete-utopia/utopia/pull/3842/commits/440e04110229c874634d93f4359093ca5d21ac7e

File: editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx (hunk 7 above)
Comment (ruggi) [reply editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx]: noice! https://github.com/concrete-utopia/utopia/pull/3842/commits/a10d54815cd6720d040cc5570f6057d7f7c04107

File: editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx (hunk 6 above)
Comment (ruggi) [reply editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx]: done in https://github.com/concrete-utopia/utopia/pull/3842/commits/4de4fe65103e0428b3f209f33331295de2f91ca7

File: editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+    return mapDropNulls((path: ElementPath): ElementOutsideVisibleArea | null => {
+      if (scaledCanvasArea == null) {
+        return null
+      }
+      const frame = framesByPathString[EP.toString(path)]
+      if (frame == null) {
+        return null
+      }
+
+      const topLeftSkew = { x: -navigatorWidth, y: 0 } as WindowPoint
```
Comment (Rheeseyb) [inline editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx:119]: Extremely minor - we have factories for constructing the various types of `Point` in `math-utils.ts`. It seems we don't have one for `WindowRectangle` though.

**Review by Rheeseyb — APPROVED** [review]
Incredible!

File: editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx (hunk 9 above)
Comment (ruggi) [reply editor/src/components/canvas/controls/elements-outside-visible-area-hooks.tsx]: maybe we can add the `WindowRectangle` helper in a separate PR

**Pull request merged**